### PR TITLE
Deliver schema baseline migration and ingestion ports

### PR DESCRIPTION
**Pull request by leynos**
Implement roadmap item 3.1.1 with a Diesel migration that creates the catalogue, descriptor, and user-state baseline tables, plus composite constraints and spatial/JSON indexes.

Introduce domain ingestion ports and Diesel outbound adapters so catalogue and descriptor materialisation writes remain behind hexagonal boundaries.

Add rstest unit migration contract checks and rstest-bdd behavioural scenarios for happy and unhappy paths, and update architecture/roadmap documentation with the final design decisions and completion state.

Keep migration execution compatible with embedded Postgres by using native PATH/POINT types and supported index definitions instead of a hard PostGIS extension dependency.

## Summary by Sourcery

Introduce a schema baseline migration for catalogue, descriptor, and user-state data, update the routes schema, and add domain ingestion ports with Diesel adapters and tests to keep materialisation behind hexagonal boundaries.

New Features:
- Add PostgreSQL schema baseline migration creating spatial POI tables, catalogue read models, and descriptor registries with composite keys and GiST/GIN indexes.
- Introduce catalogue and descriptor ingestion domain ports with fixture implementations for use by ingestion jobs.
- Provide Diesel-backed catalogue and descriptor ingestion repositories for PostgreSQL, exposed via the outbound persistence module.

Enhancements:
- Align the routes table and Diesel models with the architecture schema by replacing request-based fields with path and generation parameter columns and relaxing the user_id requirement.
- Extend Diesel schema definitions and models to cover new catalogue, descriptor, and interest theme tables while keeping geometric columns Diesel-compatible.
- Update backend architecture and roadmap documentation to record final design decisions and mark schema baseline item 3.1.1 as completed.

Tests:
- Add rstest unit checks that assert the baseline migration defines required tables, indexes, and constraints and that the down migration restores route compatibility columns.
- Add rstest-bdd behavioural tests and feature specifications that validate baseline tables exist, spatial/JSON indexes are present, and composite-key uniqueness is enforced under embedded Postgres.
- Adjust existing integration tests to work with the updated routes schema using path and generation_params columns.

### CODE REVIEW

File: backend/migrations/2026-02-06-012424_schema_baseline_catalogue_descriptor_user_state/up.sql [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+DROP TRIGGER IF EXISTS update_routes_updated_at ON routes;
+DROP INDEX IF EXISTS idx_routes_request_id;
+
+ALTER TABLE routes
+    DROP COLUMN IF EXISTS request_id,
+    DROP COLUMN IF EXISTS plan_snapshot,
+    DROP COLUMN IF EXISTS updated_at,
+    ALTER COLUMN user_id DROP NOT NULL,
+    ADD COLUMN IF NOT EXISTS path PATH NOT NULL DEFAULT '((0,0),(0,0))'::path,
+    ADD COLUMN IF NOT EXISTS generation_params JSONB NOT NULL DEFAULT '{}'::jsonb;
```
[the remark this answers is not in the document]
Comment (leynos) [reply backend/migrations/2026-02-06-012424_schema_baseline_catalogue_descriptor_user_state/up.sql]: @coderabbitai the engineer writes;

> Not technically executable in this environment: attempts to create a GiST index directly on routes.path (and expression variants) failed (no default operator class / missing function errors). To preserve migration correctness, I removed the unmatched down-migration drop and kept the migration operable.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- The Diesel catalogue/descriptor ingestion repositories perform one insert per record in a loop; consider using batched inserts with `values(&records_as_rows)` where possible to reduce round-trips and improve ingestion performance.
- The pool and Diesel error mapping logic (`map_pool_error`/`map_diesel_error`) is duplicated between the catalogue and descriptor ingestion adapters; extracting shared helpers (or a small trait) in the persistence module would reduce repetition and keep error handling consistent.
- The down migration drops `idx_routes_path_gist`, but the up migration never creates this index; either add the corresponding index creation in `up.sql` or remove the drop to keep the migration pair symmetric and self-documenting.

## Individual Comments

### Comment 1
<location> `backend/src/outbound/persistence/diesel_catalogue_ingestion_repository.rs:54-63` </location>
<code_context>
+
+#[async_trait]
+impl CatalogueIngestionRepository for FixtureCatalogueIngestionRepository {
+    async fn upsert_route_categories(
+        &self,
+        _records: &[RouteCategoryIngestion],
+    ) -> Result<(), CatalogueIngestionRepositoryError> {
+        Ok(())
+    }
+
</code_context>

<issue_to_address>
**suggestion (performance):** Consider batching inserts/updates per call instead of one statement per record

Each `upsert_*` method issues one `INSERT .. ON CONFLICT` per record, which will cause many DB round-trips for larger ingestion batches. Consider using Diesel’s multi-row insert support (e.g. `insert_into(...).values(&rows)` with `rows: Vec<NewXRow>`) so the whole batch is upserted in a single statement, ideally within a transaction for clearer all-or-nothing semantics.

Suggested implementation:

```rust
#[async_trait]
impl CatalogueIngestionRepository for DieselCatalogueIngestionRepository {

```

1. The `SEARCH` block assumes the existing `upsert_route_categories` used a `for record in records` loop with per-record `insert_into(...).on_conflict(...).do_update()` calls. If your current implementation differs, adjust the `SEARCH` section to match your actual function body so the replacement applies cleanly.
2. Replace `self.pool` with the actual async connection pool field used elsewhere in this repository (e.g. `self.connection_pool`) so it compiles consistently with the rest of the file.
3. Replace `route_categories::table`, `route_categories::id`, and the `use crate::schema::route_categories::dsl::*;` import with whatever schema/table/PK identifiers you actually use for route categories (look at the other `upsert_*` implementations for the correct schema module and primary key field).
4. The `.values(records)` and `.set(records)` calls assume that `RouteCategoryIngestion` implements the appropriate Diesel `Insertable` / `AsChangeset` traits against the `route_categories` table and that using the slice directly is already the pattern in the other `upsert_*` methods. If those methods build explicit `New*` structs, you should instead:
   - Map `records` into a `Vec<NewRouteCategoryRow>` (or equivalent insertable struct), and
   - Use `.values(&new_rows)` for the insert, and `.set(&new_rows)` or an explicit tuple of columns for the update.
5. To make all upsert methods consistent with this batching behavior, apply the same pattern (single `insert_into(...).values(records)` inside `conn.transaction`) to the other `upsert_*` methods in this impl that currently loop over `records` and issue one statement per record.
</issue_to_address>

### Comment 2
<location> `backend/src/outbound/persistence/diesel_catalogue_ingestion_repository.rs:58-67` </location>
<code_context>
+                    &self,
+                    records: &[$ingestion_type],
+                ) -> Result<(), DescriptorIngestionRepositoryError> {
+                    let mut conn = self.pool.get().await.map_err(map_pool_error)?;
+                    for record in records {
+                        let row = $row_type::from(record);
+                        diesel::insert_into($table::table)
+                            .values(&row)
+                            .on_conflict($table::id)
+                            .do_update()
+                            .set((
+                                $(
+                                    $table::$field.eq(excluded($table::$field)),
+                                )+
+                            ))
+                            .execute(&mut conn)
+                            .await
+                            .map_err(map_diesel_error)?;
+                    }
+                    Ok(())
+                }
+            )+
</code_context>

<issue_to_address>
**suggestion (bug_risk):** Using a transaction around each batch would improve consistency on partial failures

Right now each row is committed independently; if a later insert/update fails (e.g. FK violation), earlier rows remain applied and the batch no longer reflects a consistent snapshot. Consider wrapping the loop in a single transaction on the connection (e.g. `conn.transaction::<_, diesel::result::Error, _>(|conn| { ... })` or the async equivalent) so that the entire batch either succeeds or rolls back together.
</issue_to_address>

### Comment 3
<location> `backend/src/outbound/persistence/diesel_descriptor_ingestion_repository.rs:64-87` </location>
<code_context>
+                    &self,
+                    records: &[$ingestion_type],
+                ) -> Result<(), DescriptorIngestionRepositoryError> {
+                    let mut conn = self.pool.get().await.map_err(map_pool_error)?;
+                    for record in records {
+                        let row = $row_type::from(record);
+                        diesel::insert_into($table::table)
+                            .values(&row)
+                            .on_conflict($table::id)
+                            .do_update()
+                            .set((
+                                $(
+                                    $table::$field.eq(excluded($table::$field)),
+                                )+
+                            ))
+                            .execute(&mut conn)
+                            .await
+                            .map_err(map_diesel_error)?;
+                    }
+                    Ok(())
+                }
+            )+
</code_context>

<issue_to_address>
**suggestion (performance):** Descriptor upsert macro also runs one query per record; batching could simplify and speed this path

The `impl_upsert_methods!` macro currently issues one `INSERT .. ON CONFLICT` per record. Since descriptor registries are often ingested in bulk, this pattern may be unnecessarily costly. Consider building a `Vec<NewXRow>` from `records` and issuing a single `insert_into($table::table).values(&rows)...` per method to cut round trips and simplify the macro-generated code, especially once transactions are involved.

```suggestion
            $(
                async fn $method_name(
                    &self,
                    records: &[$ingestion_type],
                ) -> Result<(), DescriptorIngestionRepositoryError> {
                    if records.is_empty() {
                        return Ok(());
                    }

                    let mut conn = self.pool.get().await.map_err(map_pool_error)?;

                    let rows: Vec<$row_type> = records
                        .iter()
                        .map(|record| $row_type::from(record))
                        .collect();

                    diesel::insert_into($table::table)
                        .values(&rows)
                        .on_conflict($table::id)
                        .do_update()
                        .set((
                            $(
                                $table::$field.eq(excluded($table::$field)),
                            )+
                        ))
                        .execute(&mut conn)
                        .await
                        .map_err(map_diesel_error)?;

                    Ok(())
                }
            )+
```
</issue_to_address>

### Comment 4
<location> `backend/src/outbound/persistence/diesel_descriptor_ingestion_repository.rs:47` </location>
<code_context>
+    DescriptorIngestionRepositoryError::query(error_message)
+}
+
+macro_rules! impl_upsert_methods {
+    (
+        impl $trait:ident for $repo:ty {
</code_context>

<issue_to_address>
**issue (complexity):** Consider removing the `impl_upsert_methods!` macro and writing the upsert methods explicitly to reduce cognitive overhead and keep the implementation straightforward.

The `impl_upsert_methods!` macro doesn’t buy much here and does increase cognitive load, especially with the `keep` block and the bespoke argument list. With only four very similar methods, inlining them as normal async fns will be simpler and still easy to maintain.

You can keep all functionality intact by:

1. Removing the macro entirely.
2. Implementing `DescriptorIngestionRepository` directly with explicit methods.
3. Optionally keeping the `From` impls (or inlining them).

For example, instead of:

```rust
macro_rules! impl_upsert_methods { /* ... */ }

impl_upsert_methods! {
    impl DescriptorIngestionRepository for DieselDescriptorIngestionRepository {
        methods: [
            (
                upsert_tags,
                TagIngestion,
                NewTagRow,
                tags,
                [slug, icon_key, localizations]
            ),
            // ...
        ],
        keep: { /* upsert_interest_themes */ }
    }
}
```

Define the trait impl explicitly:

```rust
#[async_trait]
impl DescriptorIngestionRepository for DieselDescriptorIngestionRepository {
    async fn upsert_tags(
        &self,
        records: &[TagIngestion],
    ) -> Result<(), DescriptorIngestionRepositoryError> {
        let mut conn = self.pool.get().await.map_err(map_pool_error)?;
        for record in records {
            let row = NewTagRow::from(record);
            diesel::insert_into(tags::table)
                .values(&row)
                .on_conflict(tags::id)
                .do_update()
                .set((
                    tags::slug.eq(excluded(tags::slug)),
                    tags::icon_key.eq(excluded(tags::icon_key)),
                    tags::localizations.eq(excluded(tags::localizations)),
                ))
                .execute(&mut conn)
                .await
                .map_err(map_diesel_error)?;
        }
        Ok(())
    }

    async fn upsert_badges(
        &self,
        records: &[BadgeIngestion],
    ) -> Result<(), DescriptorIngestionRepositoryError> {
        let mut conn = self.pool.get().await.map_err(map_pool_error)?;
        for record in records {
            let row = NewBadgeRow::from(record);
            diesel::insert_into(badges::table)
                .values(&row)
                .on_conflict(badges::id)
                .do_update()
                .set((
                    badges::slug.eq(excluded(badges::slug)),
                    badges::icon_key.eq(excluded(badges::icon_key)),
                    badges::localizations.eq(excluded(badges::localizations)),
                ))
                .execute(&mut conn)
                .await
                .map_err(map_diesel_error)?;
        }
        Ok(())
    }

    async fn upsert_safety_toggles(
        &self,
        records: &[SafetyToggleIngestion],
    ) -> Result<(), DescriptorIngestionRepositoryError> {
        let mut conn = self.pool.get().await.map_err(map_pool_error)?;
        for record in records {
            let row = NewSafetyToggleRow::from(record);
            diesel::insert_into(safety_toggles::table)
                .values(&row)
                .on_conflict(safety_toggles::id)
                .do_update()
                .set((
                    safety_toggles::slug.eq(excluded(safety_toggles::slug)),
                    safety_toggles::icon_key.eq(excluded(safety_toggles::icon_key)),
                    safety_toggles::localizations.eq(excluded(safety_toggles::localizations)),
                ))
                .execute(&mut conn)
                .await
                .map_err(map_diesel_error)?;
        }
        Ok(())
    }

    async fn upsert_safety_presets(
        &self,
        records: &[SafetyPresetIngestion],
    ) -> Result<(), DescriptorIngestionRepositoryError> {
        let mut conn = self.pool.get().await.map_err(map_pool_error)?;
        for record in records {
            let row = NewSafetyPresetRow::from(record);
            diesel::insert_into(safety_presets::table)
                .values(&row)
                .on_conflict(safety_presets::id)
                .do_update()
                .set((
                    safety_presets::slug.eq(excluded(safety_presets::slug)),
                    safety_presets::icon_key.eq(excluded(safety_presets::icon_key)),
                    safety_presets::localizations.eq(excluded(safety_presets::localizations)),
                    safety_presets::safety_toggle_ids.eq(excluded(safety_presets::safety_toggle_ids)),
                ))
                .execute(&mut conn)
                .await
                .map_err(map_diesel_error)?;
        }
        Ok(())
    }

    async fn upsert_interest_themes(
        &self,
        records: &[InterestThemeIngestion],
    ) -> Result<(), DescriptorIngestionRepositoryError> {
        let mut conn = self.pool.get().await.map_err(map_pool_error)?;
        for record in records {
            let row = NewInterestThemeRow {
                id: record.id,
                name: record.name.as_str(),
                description: record.description.as_deref(),
            };
            diesel::insert_into(interest_themes::table)
                .values(&row)
                .on_conflict(interest_themes::id)
                .do_update()
                .set((
                    interest_themes::name.eq(excluded(interest_themes::name)),
                    interest_themes::description.eq(excluded(interest_themes::description)),
                ))
                .execute(&mut conn)
                .await
                .map_err(map_diesel_error)?;
        }
        Ok(())
    }
}
```

If you want a small DRY helper without a macro, you can still keep it local and explicit:

```rust
async fn upsert_each<T, R, F, Tab, Id, Set>(
    pool: &DbPool,
    records: &[T],
    map: F,
    table: Tab,
    id: Id,
    set: Set,
) -> Result<(), DescriptorIngestionRepositoryError>
where
    F: Fn(&T) -> R,
    R: Send + Sync,
    Tab: diesel::Table,
    Id: diesel::Expression,
    Set: diesel::Expression,
{
    let mut conn = pool.get().await.map_err(map_pool_error)?;
    for record in records {
        let row = map(record);
        diesel::insert_into(table)
            .values(&row)
            .on_conflict(id)
            .do_update()
            .set(set.clone())
            .execute(&mut conn)
            .await
            .map_err(map_diesel_error)?;
    }
    Ok(())
}
```

…but given there are just five methods, the fully inlined version above is likely the lowest-complexity approach.
</issue_to_address>

### Comment 5
<location> `docs/execplans/backend-3-1-1-diesel-migrations.md:3` </location>
<code_context>
+# Deliver Diesel schema baseline for roadmap 3.1.1
+
+This ExecPlan is a living document. The sections `Constraints`, `Tolerances`,
+`Risks`, `Progress`, `Surprises & Discoveries`, `Decision Log`, and
+`Outcomes & Retrospective` must be kept up to date as work proceeds.
</code_context>

<issue_to_address>
**suggestion (review_instructions):** "ExecPlan" is introduced as an acronym without being defined on first use.

Consider expanding this to something like "Execution Plan (ExecPlan)" on first use so that the acronym is explicitly defined, in line with the documentation guidelines.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 6
<location> `docs/execplans/backend-3-1-1-diesel-migrations.md:11` </location>
<code_context>
+
+## Purpose / Big Picture
+
+Roadmap item 3.1.1 requires a concrete Diesel migration baseline that
+materialises the backend architecture schema for catalogue, descriptor, and
+user-state persistence. After completion, the backend can apply the baseline
</code_context>

<issue_to_address>
**suggestion (review_instructions):** GiST/GIN are used later in this sentence without being defined on first use.

To follow the acronym guideline, please expand GiST and GIN the first time they appear (for example, "GiST (Generalised Search Tree) / GIN (Generalised Inverted Index) indexes") before using the acronyms alone.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 7
<location> `docs/execplans/backend-3-1-1-diesel-migrations.md:54` </location>
<code_context>
+  with embedded Postgres capabilities.
+  Severity: high
+  Likelihood: medium
+  Mitigation: use native `PATH`/`POINT` types plus supported GiST/GIN indexes
+  in baseline migration, and document the decision.
+
</code_context>

<issue_to_address>
**suggestion (review_instructions):** This line continues to use GiST/GIN acronyms without a prior expansion in the document.

After expanding GiST/GIN on first use earlier in the document, this line can remain as-is; otherwise, it currently assumes the reader already knows these acronyms, which conflicts with the guideline.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 8
<location> `docs/execplans/backend-3-1-1-diesel-migrations.md:98` </location>
<code_context>
+- Decision: keep schema baseline compatible with embedded Postgres by using
+  native `PATH`/`POINT` types and supported GiST/GIN indexes instead of a hard
+  PostGIS extension dependency.
+  Rationale: project quality gates require local test execution via
+  `pg-embedded-setup-unpriv`, and that environment does not bundle PostGIS.
+  Date/Author: 2026-02-06 / Codex.
</code_context>

<issue_to_address>
**suggestion (review_instructions):** "PostGIS" appears in this section without being defined as an acronym on first use.

Please expand PostGIS on first mention (for example, "PostGIS (PostgreSQL Geographic Information System extension)") so the acronym is explicitly defined before being used on its own.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 9
<location> `docs/execplans/backend-3-1-1-diesel-migrations.md:37` </location>
<code_context>
+
+## Tolerances (Exception Triggers)
+
+- Scope: if work exceeds 20 files or 1,200 net LOC, pause and reassess.
+- API drift: if HTTP contract changes are required, stop and escalate.
+- Dependencies: no new crates unless unavoidable.
</code_context>

<issue_to_address>
**suggestion (review_instructions):** "LOC" is used without being expanded on first use.

Consider expanding this to something like "1,200 net lines of code (LOC)" the first time it appears to comply with the acronym rule.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 10
<location> `docs/execplans/backend-3-1-1-diesel-migrations.md:38` </location>
<code_context>
+## Tolerances (Exception Triggers)
+
+- Scope: if work exceeds 20 files or 1,200 net LOC, pause and reassess.
+- API drift: if HTTP contract changes are required, stop and escalate.
+- Dependencies: no new crates unless unavoidable.
+- Validation retries: if a gate fails 3 times without clear root cause, stop
</code_context>

<issue_to_address>
**suggestion (review_instructions):** "API" is used without an initial expansion to "Application Programming Interface".

To follow the guideline, please expand API on first use in the document (for example, "Application Programming Interface (API) drift").

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 11
<location> `docs/execplans/backend-3-1-1-diesel-migrations.md:154` </location>
<code_context>
+
+## Idempotence and Recovery
+
+Migration is reversible via `down.sql`. Template DB strategy allows rerunning
+integration/behaviour tests without persistent state conflicts.
+
</code_context>

<issue_to_address>
**suggestion (review_instructions):** "DB" is used as an acronym for "database" without being defined on first use.

Please expand "DB" on first use (for example, "Template database (DB) strategy") so that the acronym is clearly introduced.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~
